### ISSUE TITLE
VIVO kernel restictions: "Operation not permitted" when executing `su` or mounting to `/system`

### ISSUE THREAD

**Issue by canyie** [topjohnwu/Magisk#5148]
Device: iQOO Z1/VIVO X70 PRO/iQOO NEO3
Android version: 10/11
Magisk version name: 0c1f6881
Magisk version code: 23016

Magisk fails to mount /system on some Vivo/iQOO devices.
```
01-01 08:17:34.721  root   650   652 E Magisk  : mount tmpfs->/system/bin failed with 1: Operation not permitted
01-01 08:17:34.721  root   650   652 E Magisk  : mount /dev/Qnh/.magisk/mirror/system/bin/AudioSetParam->/system/bin/AudioSetParam failed with 1: Operation not permitted
01-01 08:17:34.721  root   650   652 E Magisk  : mount /dev/Qnh/.magisk/mirror/system/bin/TEST_IServiceManager->/system/bin/TEST_IServiceManager failed with 1: Operation not permitted
01-01 08:17:34.721  root   650   652 E Magisk  : mount /dev/Qnh/.magisk/mirror/system/bin/abb->/system/bin/abb failed with 1: Operation not permitted
```
According to a bug report from a "Vivo internal staff" who has a device running userdebug build, `adb root` and manually mount /system works but fails when mounting in Magisk superuser environment. After debugging with him, we found that the device only allows programs in /system to mount the system partition (move binary that will mount /system to other partitions, and it fails even in adb root environment). I believe that the device has an allowlist and refuses to mount key partitions for programs if their partition is not in the allowlist.

I think the following two solutions may fix it:
1. Reimplement magic mounting logic in rc script language and add it to init.rc.
2. Inject Magisk binaries to /system in magiskinit.

I'm unsure when this problem will be solved, so I opened an issue to track the relevant information.

---

Edit: even successfully mount `su` to `/system/bin/su`, executing `su` will throw an operation not permitted error.

**Comment by yujincheng08** [comment]
`/system/bin/su` cannot be executed by other processes so there may be kernel restriction. This issue is hard to fix if no kernel source is available.

**Comment by yujincheng08** [comment]
Kernel sources of some vivo devices, but unfortunately, there are no ones of relative devices

https://opensource.vivo.com/Project

**Comment by ShazuxD** [comment]
Hey how you unlocked the bootloader of vivo phones?

**Comment by jnelson89** [comment]
> Update: Kernel source code of iQOO Z1 is currently available!

Here are my Device Vivo V5 mt6750 kernel sources which can be helpful for you to fix my problem
https://github.com/Dawn-Blossoms/android_kernel_oppo_mt6750
https://github.com/Power535/android_kernel_common_MT6763

**Comment by canyie** [comment]
Found something (Chinese) about Vivo's unqine behavior: https://bbs.pediy.com/thread-268165.htm

**Comment by jnelson89** [comment]
> Found something (Chinese) about Vivo's unqine behavior: https://bbs.pediy.com/thread-268165.htm
> 
> This points out adding `mount` to init.rc can work: https://bbs.pediy.com/thread-199415.htm

Can you make a build, to fix this?

**Comment by aIecxs** [comment]
consider Magisk is systemless-root method (and probably will stay)

**Comment by ASIMKHAN222** [comment]
> Found something (Chinese) about Vivo's unqine behavior: https://bbs.pediy.com/thread-268165.htm
> 
> This points out adding `mount` to init.rc can work: https://bbs.pediy.com/thread-199415.htm

please Guide which Tools we use to unpack or modify System.img for grant root Access

**Comment by Fox2Code** [comment]
I have though about it a bit, the problem of this issue is Magisk is not marked as a "system process" because it register itself as a service.

Maybe the `init` system is removing permission from the service processes unless a specific flag is given. 

`init` calling the `mount`/`magisk` applet calling the `mount` syscall should in no way be different of each other.

Maybe a solution would be to make `magiskinit` sleep in the background until it need to launch `magisk` itself.

This is to avoid using `init.rc` services at all cost to not get an unprivileged context from the `init` process.

**Comment by canyie** [comment]
1. adb shell magisk su -c mount xxx /system # Operation not permitted
2. adb root && adb shell mount xxx /system # OK
3. adb root && adb shell sh -c mount xxx /system # Operation not permitted
4. adb shell su # Operation not permitted
5. adb shell magisk su # OK
6. adb shell cp /sbin/su /sbin/suu && adb shell /sbin/suu # OK
7. adb root && adb shell cp /system/bin/mount /sbin && adb shell /sbin/mount # Operation not permitted

According these tests, I think the kernel:
1. Prevent binary named "su" from being executed
2. Prevent mounting to /system if the process's grandparent process is not init (in other words, the process is not created by a init's direct child process)
3. Prevent mounting to /system if the program path is not in allowlist

If so, there must be some weird restrictions in their kernel.

**Comment by canyie** [comment]
Yes but we still have another issue (apps cannot execute any binary named "su") so I tend to patch the kernel, but I cannot find anything about this weird behavior in their kernel source code. I will upload the source code to github.

**Comment by Fox2Code** [comment]
Well, making `libsu` prefer using `magisk su` rather than `su` could be a workaround.
Sure apps requiring `su` would not work, but at least apps using `libsu` could still get access to root.

**Comment by aIecxs** [comment]
does it affect symlinks?

**Comment by Fox2Code** [comment]
@aIecxs actually `su` is already a symlink to `magisk`, so yes it does affect symlinks.

**Comment by canyie** [comment]
Uploaded the kernel source code to https://github.com/canyie/iQOO-Z1-kernel

**Comment by Parafinchik** [comment]
+


вс, 20 февр. 2022 г., 06:02 jnelson89 ***@***.***>:

**Comment by canyie** [comment]
Trying to execute any binary named "su" in Twoyi (an android container that interprets syscalls and run another Android system on an android device) or even in TWRP will also throw an operation not permitted error. So this may be a hardware restriction, wanna know why they did strange things to their devices 🤔

**Comment by Fox2Code** [comment]
@canyie it's really common for most TWRP recoveries to use stock kernel to work properly on a device.
I still think focusing on analyzing the kernel source code to resolve the issue is the best idea.
I already made a pull request to `libsu` to help workaround the `su` issue. -> https://github.com/topjohnwu/libsu/pull/104

**Comment by Fox2Code** [comment]
The "Prevent mounting to /system if the program path is not in allowlist", maybe it's possible to get magisk daemon to be in the allow list by using `LD_PRELOAD` on a program on the allow list, the `mount` and `sh` applets seems to be on the allow list.

**Comment by yujincheng08** [comment]
@Fox2Code you got a vivo device? are you sure `mount` is allowed? as far as we tested, `mount` still produces `EPERM`.

but the mount issue can be easily bypassed. i hv done one. the core issue is `su`. we read the kernel source and nothing helped. so I assume there're some hardware constraints.

**Comment by Fox2Code** [comment]
@yujincheng08 nah, I'm just theorizing, if I didn't had so many problems in my life and the money I would probably just buy one to try to fix the bug, unfortunately I need to take care of my life first.

**Comment by canyie** [comment]
(Unconfirmed) A Vivo S5 (Android 10) user reports device refuses to boot if `su` does exist in `PATH`. No logs were provided.

**Comment by romanovj** [comment]
people reporting working magisk on stock rom
vivo xfold 24309
vivo x70 pro plus 24306

I have vivo y31 2021, but magisk 24300 can only grant root on LOS 19.1 GSI bvS with removed phh-su. It's can't mount modules, can't mount -o remount,rw /

But adb root can do this. To test zygisk and modules I manualy mounting then during boot process with adb root, adb shell stop etc

**Comment by qiurigao** [comment]
@topjohnwu

**Comment by qiurigao** [comment]
Why is there no su limit when uid=0?
为啥uid=0的时候没有su限制呢

**Comment by qiurigao** [comment]
> 是的，但我们还有另一个问题（应用程序无法执行任何名为“su”的二进制文件）所以我倾向于修补内核，但我在他们的内核源代码中找不到任何关于这种奇怪行为的信息。我会把源代码上传到github。

不，他只是限制非root用户，如果你切换到adb root 执行su不会收到阻碍

**Comment by canyie** [comment]
> > 是的，但我们还有另一个问题（应用程序无法执行任何名为“su”的二进制文件）所以我倾向于修补内核，但我在他们的内核源代码中找不到任何关于这种奇怪行为的信息。我会把源代码上传到github。
> 
> 不，他只是限制非root用户，如果你切换到adb root 执行su不会收到阻碍

但问题是，应用执行不了 su 也就切换不到 root 用户。adb shell 其实是可以执行 su 的，但是 app 就会被拒绝

**Comment by qiurigao** [comment]
> > > 是的，但我们还有另一个问题（应用程序无法执行任何名为“su”的二进制文件）所以我倾向于修补内核，但我在他们的内核源代码中找不到任何关于这种奇怪行为的信息。我会把源代码上传到github。
> > 
> > 
> > 不，他只是限制非root用户，如果你切换到adb root 执行su不会收到阻碍
> 
> 但问题是，应用执行不了 su 也就切换不到 root 用户。adb shell 其实是可以执行 su 的，但是 app 就会被拒绝

shell用户(uid=2000)我这边测试的是无法运行文件名为su的任何程序，放切换到root环境(包括TWRP)时，su不会收到限制。
因此我猜想可能是vendor供应商某个.so文件的作用，因为我曾经编译过los，当系统在开机第二屏幕时，su没有限制(此时blob文件未加载)。等进入系统后，su就被限制了。另一个猜想是kernel限制，可是我通过IDA 分析以后，并且修改了所有可能的su字符串，此限制依旧。所以我人为内核的限制可能性较小，如果有可能，有没有办法在twrp环境中使adb shell的权限为uid2000,然后尝试su限制状态

**Comment by qiurigao** [comment]
> > > 是的，但我们还有另一个问题（应用程序无法执行任何名为“su”的二进制文件）所以我倾向于修补内核，但我在他们的内核源代码中找不到任何关于这种奇怪行为的信息。我会把源代码上传到github。
> > 
> > 
> > 不，他只是限制非root用户，如果你切换到adb root 执行su不会收到阻碍
> 
> 但问题是，应用执行不了 su 也就切换不到 root 用户。adb shell 其实是可以执行 su 的，但是 app 就会被拒绝

我只是在寻找su限制的具体源头，然后通过eflpatch修改，已达到绕过su限制的目的

**Comment by canyie** [comment]
So according to #5971 there are some devices don't even allow magiskinit to mount something to /system... wtf?!

**Comment by qiurigao** [comment]
> 所以根据#5971有一些设备甚至不允许 magiskinit 将某些东西挂载到 /system... wtf？！

这可能是另一个限制，因为vivo设备运行TWRP环境时，甚至不允许刷写system分区，似乎和su的限制相同，然而官方rec居然没有限制，我猜测是做了某种修改

**Comment by yujincheng08** [comment]
哪来那么多猜测，直接编译一个 kernel 测试不就好。

**Comment by qiurigao** [comment]
> 来猜猜，直接编译一个内核不就好了。

前提是拥有完整的内核源代码，别指望他们的开源，那些都是半成品

**Comment by canyie** [comment]
iQOO Z1 的内核源码：https://github.com/canyie/iqoo-z1-kernel ，不知道能不能编译，我是懒得搞了

**Comment by qiurigao** [comment]
> iQOO Z1 的内核内核：https://github.com/canyie/iqoo-z1-kernel，不知道能编译，我是懒得搞了

我安装了lineage-19.1-20220613-UNOFFICIAL-arm64_bvS的GSI后，发现su完全没有限制,adb 和应用均可以执行su,该版本为userdebug,难道是内核限制user版本无法执行su？

**Comment by canyie** [comment]
@qiurigao 有人用 vivo 官方的 userdebug 测试过了，adb root 可以运行，adb shell 也可以执行 su，但是 app 还是受到限制

**Comment by qiurigao** [comment]
> @qiurigao 有人用 vivo 官方的 userdebug 测试过了，adb root 可以运行，adb shell 也可以执行 su，但是 app 还是受到限制

这个我也测试过刷入boot-debug.img就有adb shell，不过仅此而已。可能和system有关
![IMG20220616013127](https://user-images.githubusercontent.com/22121479/173992250-509ab71a-cb8e-4a1f-a27f-10237ef067fe.jpg)

**Comment by olbb** [comment]
It looks like the kernel is restricting app access to su.
We have a way to make adb shell have root privileges.
Is there a way to make magisk work properly,
 by manually modifying system files or by adding some magisk files to the system?

**Comment by yujincheng08** [comment]
@olbb compile customized kernel

**Comment by olbb** [comment]
Thank U for reply.
I'm  afraid I can't because my device's kernel code is not opened.

**Comment by olbb** [comment]
``` cpp
#define CONFIG_MOUNT_RESTRICT
	/*#define PASS_MOUNT_RESTRICT*/
	#define CONFIG_MOUNT_BIND_MOVE_RESTRICT /*限制 mount -o bind/move*/
```
https://github.com/vivo-source-mirror/Y53s/blob/2b1b1f054ccce7ed551137f0eb1d0cce6976f2aa/include/linux/vrr.hpp

```cpp
#if defined(CONFIG_RS_CHECK_FUNC) && defined(CONFIG_MOUNT_RESTRICT)
	#define	VIVO_DEFINE_FC_DO_MOUNT_BEGIN_STUB	\
		RS_DEFINE_FC_BEGIN_PUB(do_mount);			\
		RS_FORWARD_FC_END_PUB(do_mount);

	#define	VIVO_DEFINE_FC_DO_MOUNT_BODY	RS_DEFINE_FC_BODY(do_mount)

	#define	VIVO_DEFINE_FC_DO_MOUNT_END_STUB	\
		RS_DEFINE_FC_END_PUB(do_mount);
#else
	#define	VIVO_DEFINE_FC_DO_MOUNT_BEGIN_STUB
	#define	VIVO_DEFINE_FC_DO_MOUNT_BODY
	#define	VIVO_DEFINE_FC_DO_MOUNT_END_STUB
#endif
```

**Comment by yujincheng08** [comment]
I think it should be a won't fix because those configs are too many, and bypassing them one by one is tricky.

**Comment by olbb** [comment]
Successed! By modify the kernel file with IDA.
But it is really difficult to find a common way to patch vivo roms. 
The address of the restriction method in different roms may be different.
![Screenshot_20220701_225219](https://user-images.githubusercontent.com/5554102/176735562-db898d88-ab06-48b4-a047-641bf491b8e8.jpg)
![Screenshot_20220701_225224](https://user-images.githubusercontent.com/5554102/176735580-4481ac7c-1825-4633-b78d-9cbbb015184d.jpg)
![Screenshot_20220701_225245](https://user-images.githubusercontent.com/5554102/176735603-334e20b1-cdd5-4bb1-8e90-b9c33504d666.jpg)

**Comment by olbb** [comment]
> > 
> 
> can you post wich things did you change?

The 'do_mount_check' method(Maybe confused in kernel)

**Comment by romanovj** [comment]
> 

@olbb 
=( no such method in /proc/kallsyms (vivo y31 2021)

**Comment by qiurigao** [comment]
> 成功了！通过使用 IDA 修改内核文件。 但是要找到一个通用的方法来修补vivo rom真的很难。 不同的rom中限制方法的地址可能不同。 <img alt="" width="218" height="468" src="https://user-images.githubusercontent.com/5554102/176735562-db898d88-ab06-48b4-a047-641bf491b8e8.jpg"> <img alt="" width="218" height="468" src="https://user-images.githubusercontent.com/5554102/176735580-4481ac7c-1825-4633-b78d-9cbbb015184d.jpg"> <img alt="" width="218" height="468" src="https://user-images.githubusercontent.com/5554102/176735603-334e20b1-cdd5-4bb1-8e90-b9c33504d666.jpg">

Can you share your modification process, maybe I can refer to it

**Comment by olbb** [comment]
Take a look at 'ssc_utils.c' in Y53s's kernel code.
``` cpp
#if 1 /*defined(CONFIG_RS_OBFUSCATED_NAME)*/
/*有可能把user版判为eng版？*/
#define RS_HIDE(name) name##_v_r_s_h_s
#else
#define RS_HIDE(name) name
#endif
```
```
#define do_mount_check RS_HIDE(_yb)
```
The do_mount_check method is confused to "_yb_v_r_s_h_s".
Maybe different in different roms. 
Find the method in ida,  and modify and let the method ret 0.

**Comment by qiurigao** [comment]
> 

关于su的限制在内核中找到了吗

**Comment by olbb** [comment]
> > 
> 
> 关于su的限制在内核中找到了吗

https://github.com/topjohnwu/Magisk/issues/5148#issuecomment-1167697477

**Comment by romanovj** [comment]
> Take a look at 'ssc_utils.c' in Y53s's kernel code.
> 
> ```c++
> #if 1 /*defined(CONFIG_RS_OBFUSCATED_NAME)*/
> /*有可能把user版判为eng版？*/
> #define RS_HIDE(name) name##_v_r_s_h_s
> #else
> #define RS_HIDE(name) name
> #endif
> ```
> 
> ```
> #define do_mount_check RS_HIDE(_yb)
> ```
> 
> The do_mount_check method is confused to "_yb_v_r_s_h_s". Maybe different in different roms. Find the method in ida, and modify and let the method ret 0.

nothing like this in /proc/kallsyms , no confiscated names at all...

can you upload yours /proc/kallsyms?

**Comment by olbb** [comment]
[kallsyms.txt](https://github.com/topjohnwu/Magisk/files/9028320/kallsyms.txt)

Your device may used a different obfuscation rule.
Try checking your device's kernel source code if available.

**Comment by romanovj** [comment]
> [kallsyms.txt](https://github.com/topjohnwu/Magisk/files/9028320/kallsyms.txt)
> 
> Your device may used a different obfuscation rule. Try checking your device's kernel source code if available.

thx, I have searched another snapdragon devices and didn't find obfuscations 

Well, I think that my situation is different. I deleted "ro" for vendor partition in fstab.default in boot ramdisk, so I have /vendor rw from the start. 

And I can execute mount -o remount,ro /vendor
but can't remount,rw 

strace -e mount mount -n -o remount,rw /vendor
mount("/dev/block/dm-1", "/vendor", 0xb400007af4edd4a8, MS_REMOUNT|MS_SILENT|MS_RELATIME, "seclabel,discard") = -1 EPERM (Operation not permitted)
mount: '/dev/block/dm-1'->'/vendor': Operation not permitted
+++ exited with 1 +++
trace -e mount mount -n -o remount,ro /vendor                             <
mount("/dev/block/dm-1", "/vendor", 0xb400007dc933db28, MS_RDONLY|MS_REMOUNT|MS_SILENT|MS_RELATIME, "seclabel,discard") = 0
+++ exited with 0 +++

Maybe it's something to do with this

long do_mount(const char *dev_name, const char __user *dir_name,
		const char *type_page, unsigned long flags, void *data_page)
{
	struct path path;
	unsigned int mnt_flags = 0, sb_flags;
	int retval = 0;

	/* Discard magic */
	if ((flags & MS_MGC_MSK) == MS_MGC_VAL)
		flags &= ~MS_MGC_MSK;

	/* Basic sanity checks */
	if (data_page)
		((char *)data_page)[PAGE_SIZE - 1] = 0;

	if (flags & MS_NOUSER)
		return -EINVAL;

	/* ... and get the mountpoint */
	retval = user_path(dir_name, &path);
	if (retval)
		return retval;

	retval = security_sb_mount(dev_name, &path,
				   type_page, flags, data_page);
	if (!retval && !may_mount())
		retval = -EPERM;
	if (!retval && (flags & SB_MANDLOCK) && !may_mandlock())
		retval = -EPERM;
	if (retval)
		goto dput_out;

	/* Default to relatime unless overriden */
	if (!(flags & MS_NOATIME))
		mnt_flags |= MNT_RELATIME;

	/* Separate the per-mountpoint flags */
	if (flags & MS_NOSUID)
		mnt_flags |= MNT_NOSUID;
	if (flags & MS_NODEV)
		mnt_flags |= MNT_NODEV;
	if (flags & MS_NOEXEC)
		mnt_flags |= MNT_NOEXEC;
	if (flags & MS_NOATIME)
		mnt_flags |= MNT_NOATIME;
	if (flags & MS_NODIRATIME)
		mnt_flags |= MNT_NODIRATIME;
	if (flags & MS_STRICTATIME)
		mnt_flags &= ~(MNT_RELATIME | MNT_NOATIME);
	if (flags & MS_RDONLY)
		mnt_flags |= MNT_READONLY;

	/* The default atime for remount is preservation */
	if ((flags & MS_REMOUNT) &&
	    ((flags & (MS_NOATIME | MS_NODIRATIME | MS_RELATIME |
		       MS_STRICTATIME)) == 0)) {
		mnt_flags &= ~MNT_ATIME_MASK;
		mnt_flags |= path.mnt->mnt_flags & MNT_ATIME_MASK;
	}

	sb_flags = flags & (SB_RDONLY |
			    SB_SYNCHRONOUS |
			    SB_MANDLOCK |
			    SB_DIRSYNC |
			    SB_SILENT |
			    SB_POSIXACL |
			    SB_LAZYTIME |
			    SB_I_VERSION);

	if (flags & MS_REMOUNT)
		retval = do_remount(&path, flags, sb_flags, mnt_flags,
				    data_page);
	else if (flags & MS_BIND)
		retval = do_loopback(&path, dev_name, flags & MS_REC);
	else if (flags & (MS_SHARED | MS_PRIVATE | MS_SLAVE | MS_UNBINDABLE))
		retval = do_change_type(&path, flags);
	else if (flags & MS_MOVE)
		retval = do_move_mount(&path, dev_name);
	else
		retval = do_new_mount(&path, type_page, sb_flags, mnt_flags,
				      dev_name, data_page);
dput_out:
	path_put(&path);
	return retval;
}

**Comment by olbb** [comment]
I think this feature might start from the origin os. 
When my device is FuntouchOS on Android 10 and doesn't have these obfuscations. 
After I upgraded to Anroid11, I found those. 
FuntouchOS may be another rule.

**Comment by qiurigao** [comment]
> 查看 Y53s 内核代码中的“ssc_utils.c”。
> 
> ```c++
> #if 1 /*defined(CONFIG_RS_OBFUSCATED_NAME)*/
> /*有可能把user版判为eng版？*/
> #define RS_HIDE(name) name##_v_r_s_h_s
> #else
> #define RS_HIDE(name) name
> #endif
> ```
> 
> ```
> #define do_mount_check RS_HIDE(_yb)
> ```
> 
> do_mount_check 方法与“_yb_v_r_s_h_s”混淆。 可能在不同的rom中有所不同。 找到ida中的方法，修改并让方法ret 0。

有没有版本找到"Operation not permitted" when executing su o去除的办法

**Comment by olbb** [comment]
> 有没有版本找到"Operation not permitted" when executing su o去除的办法

hook the 'can_exec_su' method @qiurigao

**Comment by NeonzGamingYT** [comment]
# If you guys still want to discuss, join telegram group
https://t.me/vivodev_discuss

**Comment by qiurigao** [comment]
> 



> `/system/bin/su`不能被其他进程执行，因此可能存在内核限制。如果没有可用的内核源，这个问题的解决非常严重。
针对vivo设备并且内核小于5的机型有两个限制需要对抗
这是两个步骤：
首先解决mount问题，让magisk正常挂载和运行。（出厂安卓11的设备无法正常mount）
其次修改su指令（这一步已经实现），在基于安卓10的vivo设备上已经验证过了。

**Comment by washo4evr** [comment]
Hi everyone,

this seems to be happening to vivo X80 pro as well
kernel source is available
could someone point me in the direction so I can patch the kernel / build a new one where this doesnt happen?
Thanks

**Comment by NeonzGamingYT** [comment]
> Hi everyone,
> 
> this seems to be happening to vivo X80 pro as well kernel source is available could someone point me in the direction so I can patch the kernel / build a new one where this doesnt happen? Thanks
> 
> edit: I found
> 
> 0000000000000000 t rootwait_setup 0000000000000000 t root_dev_setup 0000000000000000 t root_delay_setup 0000000000000000 t root_data_setup 0000000000000000 t readwrite 0000000000000000 t readonly 0000000000000000 t load_ramdisk 0000000000000000 t fs_names_setup 0000000000000000 T mount_block_root 0000000000000000 t get_fs_names 0000000000000000 t do_mount_root 0000000000000000 T mount_root 0000000000000000 t create_dev 0000000000000000 T prepare_namespace 0000000000000000 T init_rootfs 0000000000000000 t ramdisk_start_setup 0000000000000000 t prompt_ramdisk
> 
> in the /proc/kallsyms seems like a good place to start \x80\common\init\do_mounts.c has 'root_dev_setup' for instance

How do you get kallsyms since there is no SU

**Comment by washo4evr** [comment]
I get su access
both from ADB, or from the termux, or total commander
I can explore all the folders for instance

but other modules like LSposed don't work

**Comment by romanovj** [comment]
> I get su access both from ADB, or from the termux, or total commander I can explore all the folders for instance
> 
> but other modules like LSposed don't work

just mount zygisk files with adb root then restart zygote

mount $(magisk --path)/.magisk/zygisk/app_process64 /system/bin/app_process64
mount $(magisk --path)/.magisk/zygisk/app_process32 /system/bin/app_process32
killall surfaceflinger

alternative way is to inject .rc script and mount zygisk in init.rc

**Comment by NeonzGamingYT** [comment]
> A iQOO 9 Pro (Android 11, kernel 5.10.66) user reported that even after he/she/they flashed a GKI kernel which was compiled from Google's source code, as soon as an app process got root permission (uid=0), the process would be killed. This behavior persisted no matter SELinux status. Tried to change SELinux policy to prevent anyone from sending SIGKILL to the root process, issue persisted. This prevented the user from using KernelSU, but Magisk and Magica worked.
> 
> What a f**king thing Vivo did!

Does magisk work with GKI kernel?

**Comment by romanovj** [comment]
this patch works, can mount /system after it
+  # patch vivo do_mount_check
+  # /system -> /syswxl
+  # python3 -c "print(hex(int.from_bytes(b'/system', 'big')^int('bdbcbbbab9b8b7', 16)))"
+  ./magiskboot hexpatch kernel 0092CFC2C9CDDDDA00 0092CFC2C9CEC0DB00
https://github.com/wuxianlin/build_magisk_vivo/blob/master/patches/Magisk/patch_vivo_do_mount_check.diff

and on userdebug gsi kernel doesn't prevent execution of su

remounting / could be done from adb root

**Comment by muhammadrafiasyddiq** [comment]
> tambalan ini berfungsi, dapat memasang / sistem setelahnya
> 
> * # patch vivo do_mount_check
> * # /sistem -> /syswxl
> * # python3 -c "print(hex(int.from_bytes(b'/system', 'big')^int('bdbcbbbab9b8b7', 16)))"
> * ./magiskboot hexpatch kernel 0092CFC2C9CDDDDA00 0092CFC2C9CEC0DB00 
>   https://github.com/wuxianlin/build_magisk_vivo/blob/master/patches/Magisk/patch_vivo_do_mount_check.diff
> 
> dan pada kernel userdebug gsi tidak mencegah eksekusi su
> 
> remounting / dapat dilakukan dari adb root

i think no kernel but firmware vendor cannot execute su  if build variant user

**Comment by romanovj** [comment]
actually there are requirements for unblocking su execution

read fs/vrr.c in https://github.com/vivo-source-mirror/Y53s

no blocking SU on LOS20 pre rooted userdebug gsi (vivo y31 2021), but it's become  blocked after renaming /system/xbin/su 

from source:

noused notrace void set_rs_s_u(void)
{
#if !(defined(RS_IS_ENG_BUILD) || defined(BBK_FOR_NET_ENTRY))
        /* /system/xbin/su must exist! */

There'are also list of files yo search
check_string g_eng_mode_files

Do to unblock SU it's should be enough to have some files

**Comment by romanovj** [comment]
> A iQOO 9 Pro (Android 11, kernel 5.10.66) user reported that even after he/she/they flashed a GKI kernel which was compiled from Google's source code, as soon as an app process got root permission (uid=0), the process would be killed. This behavior persisted no matter SELinux status. Tried to change SELinux policy to prevent anyone from sending SIGKILL to the root process, issue persisted. This prevented the user from using KernelSU, but Magisk and Magica worked.
> 
> What a f**king thing Vivo did!

probably killed by kernel itself, in my kernel there are 
task_in_vivo_white_list
task_in_vivo_critical_list

which is called by "check_hung_task"

I don't if it's related

**Comment by muhammadrafiasyddiq** [comment]
> sebenarnya ada persyaratan untuk membuka blokir eksekusi su
> 
> baca fs/vrr.c di https://github.com/vivo-source-mirror/Y53s
> 
> tidak ada pemblokiran SU pada LOS20 pra-rooted userdebug gsi (vivo y31 2021), tetapi menjadi diblokir setelah mengganti nama /system/xbin/su
> 
> dari sumber:
> 
> noused notrace void set_rs_s_u(void) { #if !(defined(RS_IS_ENG_BUILD) || defined(BBK_FOR_NET_ENTRY)) /* /system/xbin/su harus ada! */
> 
> Ada juga daftar file yang Anda cari check_string g_eng_mode_files
> 
> untuk membuka blokir SU cukup memiliki beberapa file di firmware atau temukan cara menambal kernel di sini
what file need , i builded kernel vivo  and su can exec and mount 
but i get bugs in source code broken

**Comment by muhammadrafiasyddiq** [comment]
> > A iQOO 9 Pro (Android 11, kernel 5.10.66) user reported that even after he/she/they flashed a GKI kernel which was compiled from Google's source code, as soon as an app process got root permission (uid=0), the process would be killed. This behavior persisted no matter SELinux status. Tried to change SELinux policy to prevent anyone from sending SIGKILL to the root process, issue persisted. This prevented the user from using KernelSU, but Magisk and Magica worked.
> > What a f**king thing Vivo did!
> 
> probably killed by kernel itself, in my kernel there are task_in_vivo_white_list task_in_vivo_critical_list
> 
> which is called by "check_hung_task"
> 
> I don't if it's related

I Rebuild kernel  Vivo Y21s  su Can exec And mount Working but
kernel have many bug because missing driver

**Comment by romanovj** [comment]
I developed patches for unblocking SU, should be one hit per patch

./magiskboot hexpatch kernel 00010034a000805221008052 20000034a000805221008052

./magiskboot hexpatch kernel e0000034a000805221008052 20000034a000805221008052

**Comment by romanovj** [comment]
I disabled module signature verification(patch) and manually found address of call_usermodehelper()

Then I compiled kernel module out-of-tree (no sources for my phone)

With simple module I become kernel

```
#include <linux/kernel.h>
#include <linux/module.h>
#include <linux/init.h>
#include <linux/kmod.h>


int (*my_call_usermodehelper)(const char *, char **, char **, int)=(int(*)(const char *, char **, char **, int ))0xffffff80080e2b38;
static int __init init_my_module(void)
{
        char* argv[] = {"/bin/busybox", "nc", "-vlp", "5556", "-e", "/bin/sh", NULL};
        static char* envp[] = {"HOME=/","TERM=linux","PATH=/sbin:/bin:/system/bin:/system_ext/bin", NULL };
        return my_call_usermodehelper(argv[0], argv, envp, UMH_WAIT_EXEC);
}

static void __exit exit_my_module(void)
{
        printk(KERN_INFO "Bye, module!\n");
}

module_init(init_my_module);
module_exit(exit_my_module);
MODULE_LICENSE("GPL");
MODULE_AUTHOR("RRR");
```

In termux

nc localhost 5556
id
uid=0(root) gid=0(root) groups=0(root) context=u:r:kernel:s0
whoami
root

Idk if it's good idea to abuse usermodehelper 🤷

**Comment by romanovj** [comment]
Successfully patched different kernel, use radare2 to find address
SU:
/x .fc.0171....0054....4039.fd.0171....0054
should  be two hits, replace first fc with fd

Mount/remount
/x 085040b9....0034....4039
one hit, replace 0034 with 0035
(if you're unlucky then use another combinations to find and patch vrr_do_mount_check_before_mount(see above) or just apply patch only for /system (see in posts above)
![Screenshot_20230527-090244_Termux](https://github.com/topjohnwu/Magisk/assets/105647092/fdeec048-0c0f-4ca5-9624-40704e6077c3)
![Screenshot_20230527-090109_Termux](https://github.com/topjohnwu/Magisk/assets/105647092/57c5e8a0-2541-43b7-be3a-d9446ec719f8)

**Comment by muhammadrafiasyddiq** [comment]
> Successfully patched different kernel, use radare2 to find address SU: /x .fc.0171....0054....4039.fd.0171....0054 should be two hits, replace first fc with fd
> 
> Mount/remount /x 085040b9....0034....4039 one hit, replace 0034 with 0035 (if you're unlucky then use another combinations to find and patch vrr_do_mount_check_before_mount(see above) or just apply patch only for /system (see in posts above) ![Screenshot_20230527-090244_Termux](https://user-images.githubusercontent.com/105647092/241361426-fdeec048-0c0f-4ca5-9624-40704e6077c3.png) ![Screenshot_20230527-090109_Termux](https://user-images.githubusercontent.com/105647092/241361432-57c5e8a0-2541-43b7-be3a-d9446ec719f8.png)

you patch get bug in my kernel

**Comment by muhammadrafiasyddiq** [comment]
> Successfully patched different kernel, use radare2 to find address SU: /x .fc.0171....0054....4039.fd.0171....0054 should be two hits, replace first fc with fd
> 
> Mount/remount /x 085040b9....0034....4039 one hit, replace 0034 with 0035 (if you're unlucky then use another combinations to find and patch vrr_do_mount_check_before_mount(see above) or just apply patch only for /system (see in posts above) ![Screenshot_20230527-090244_Termux](https://user-images.githubusercontent.com/105647092/241361426-fdeec048-0c0f-4ca5-9624-40704e6077c3.png) ![Screenshot_20230527-090109_Termux](https://user-images.githubusercontent.com/105647092/241361432-57c5e8a0-2541-43b7-be3a-d9446ec719f8.png)

faced bug after you patch

**Comment by romanovj** [comment]
> > Successfully patched different kernel, use radare2 to find address SU: /x .fc.0171....0054....4039.fd.0171....0054 should be two hits, replace first fc with fd
> > Mount/remount /x 085040b9....0034....4039 one hit, replace 0034 with 0035 (if you're unlucky then use another combinations to find and patch vrr_do_mount_check_before_mount(see above) or just apply patch only for /system (see in posts above) 
> 
> faced bug after you patch

ok!

**Comment by muhammadrafiasyddiq** [comment]
> > > Successfully patched different kernel, use radare2 to find address SU: /x .fc.0171....0054....4039.fd.0171....0054 should be two hits, replace first fc with fd
> > > Mount/remount /x 085040b9....0034....4039 one hit, replace 0034 with 0035 (if you're unlucky then use another combinations to find and patch vrr_do_mount_check_before_mount(see above) or just apply patch only for /system (see in posts above)
> > 
> > 
> > faced bug after you patch
> 
> ok!
gsi or dsu cannot boot / decrypt

**Comment by romanovj** [comment]
> > > > Successfully patched different kernel, use radare2 to find address SU: /x .fc.0171....0054....4039.fd.0171....0054 should be two hits, replace first fc with fd
> > > > Mount/remount /x 085040b9....0034....4039 one hit, replace 0034 with 0035 (if you're unlucky then use another combinations to find and patch vrr_do_mount_check_before_mount(see above) or just apply patch only for /system (see in posts above)
> > > 
> > > 
> > > faced bug after you patch
> > 
> > 
> > ok!
> > gsi or dsu cannot boot / decrypt

did you able to do it before?

**Comment by muhammadrafiasyddiq** [comment]
> > > > > Berhasil menambal kernel yang berbeda, gunakan radare2 untuk menemukan alamat SU: /x .fc.0171....0054....4039.fd.0171....0054 harus dua hit, ganti fc pertama dengan fd 
> > > > > Mount/remount /x 085040b9....0034....4039 satu pukulan, ganti 0034 dengan 0035 (jika Anda tidak beruntung maka gunakan kombinasi lain untuk menemukan dan menambal vrr_do_mount_check_before_mount (lihat di atas) atau cukup terapkan tambalan hanya untuk / sistem (lihat di posting di atas)
> > > > 
> > > > 
> > > > menghadapi bug setelah Anda menambal
> > > 
> > > 
> > > Oke! 
> > > gsi atau dsu tidak bisa boot / mendekripsi
> 
> apakah kamu bisa melakukannya sebelumnya?
yess gsi google and stock work fine before now only bootloop

**Comment by romanovj** [comment]
> > > > > > Berhasil menambal kernel yang berbeda, gunakan radare2 untuk menemukan alamat SU: /x .fc.0171....0054....4039.fd.0171....0054 harus dua hit, ganti fc pertama dengan fd
> > > > > > Mount/remount /x 085040b9....0034....4039 satu pukulan, ganti 0034 dengan 0035 (jika Anda tidak beruntung maka gunakan kombinasi lain untuk menemukan dan menambal vrr_do_mount_check_before_mount (lihat di atas) atau cukup terapkan tambalan hanya untuk / sistem (lihat di posting di atas)
> > > > > 
> > > > > 
> > > > > menghadapi bug setelah Anda menambal
> > > > 
> > > > 
> > > > Oke!
> > > > gsi atau dsu tidak bisa boot / mendekripsi
> > 
> > 
> > apakah kamu bisa melakukannya sebelumnya?
> > yess gsi google and stock work fine before now only bootloop

so my patches worked on stock rom, but bootloops on gsi, right?

magisk SUU worked on gsi?

**Comment by muhammadrafiasyddiq** [comment]
> > > > > > > Berhasil menambal kernel yang berbeda, gunakan radare2 untuk menemukan alamat SU: /x .fc.0171....0054....4039.fd.0171....0054 harus dua hit, ganti fc pertama dengan fd
> > > > > > > Mount/remount /x 085040b9....0034....4039 satu pukulan, ganti 0034 dengan 0035 (jika Anda tidak beruntung maka gunakan kombinasi lain untuk menemukan dan menambal vrr_do_mount_check_before_mount (lihat di atas) atau cukup terapkan tambalan hanya untuk / sistem (lihat di posting di atas)
> > > > > > 
> > > > > > 
> > > > > > menghadapi bug setelah Anda menambal
> > > > > 
> > > > > 
> > > > > Oke!
> > > > > gsi atau dsu tidak bisa boot / mendekripsi
> > > 
> > > 
> > > apakah kamu bisa melakukannya sebelumnya?
> > > yess gsi google and stock work fine before now only bootloop
> 
> so my patches worked on stock rom, but bootloops on gsi, right?
> 
> magisk SUU worked on gsi?

yes stock rom fine 
but dsu slot bootloop all rom

**Comment by romanovj** [comment]
> > > > > > > > Berhasil menambal kernel yang berbeda, gunakan radare2 untuk menemukan alamat SU: /x .fc.0171....0054....4039.fd.0171....0054 harus dua hit, ganti fc pertama dengan fd
> > > > > > > > Mount/remount /x 085040b9....0034....4039 satu pukulan, ganti 0034 dengan 0035 (jika Anda tidak beruntung maka gunakan kombinasi lain untuk menemukan dan menambal vrr_do_mount_check_before_mount (lihat di atas) atau cukup terapkan tambalan hanya untuk / sistem (lihat di posting di atas)
> > > > > > > 
> > > > > > > 
> > > > > > > menghadapi bug setelah Anda menambal
> > > > > > 
> > > > > > 
> > > > > > Oke!
> > > > > > gsi atau dsu tidak bisa boot / mendekripsi
> > > > 
> > > > 
> > > > apakah kamu bisa melakukannya sebelumnya?
> > > > yess gsi google and stock work fine before now only bootloop
> > 
> > 
> > so my patches worked on stock rom, but bootloops on gsi, right?
> > magisk SUU worked on gsi?
> 
> yes stock rom fine but dsu slot bootloop all rom

just flash gsi into system partition xD

**Comment by muhammadrafiasyddiq** [comment]
> > > > > > > > > Berhasil menambal kernel yang berbeda, gunakan radare2 untuk menemukan alamat SU: /x .fc.0171....0054....4039.fd.0171....0054 harus dua hit, ganti fc pertama dengan fd
> > > > > > > > > Mount/remount /x 085040b9....0034....4039 satu pukulan, ganti 0034 dengan 0035 (jika Anda tidak beruntung maka gunakan kombinasi lain untuk menemukan dan menambal vrr_do_mount_check_before_mount (lihat di atas) atau cukup terapkan tambalan hanya untuk / sistem (lihat di posting di atas)
> > > > > > > > 
> > > > > > > > 
> > > > > > > > menghadapi bug setelah Anda menambal
> > > > > > > 
> > > > > > > 
> > > > > > > Oke!
> > > > > > > gsi atau dsu tidak bisa boot / mendekripsi
> > > > > 
> > > > > 
> > > > > apakah kamu bisa melakukannya sebelumnya?
> > > > > yess gsi google and stock work fine before now only bootloop
> > > 
> > > 
> > > so my patches worked on stock rom, but bootloops on gsi, right?
> > > magisk SUU worked on gsi?
> > 
> > 
> > yes stock rom fine but dsu slot bootloop all rom
> 
> just flash gsi into system partition xD

i don't know why keymaster cannot decrypt userdata

**Comment by romanovj** [comment]
> > > > > > > > > > Berhasil menambal kernel yang berbeda, gunakan radare2 untuk menemukan alamat SU: /x .fc.0171....0054....4039.fd.0171....0054 harus dua hit, ganti fc pertama dengan fd
> > > > > > > > > > Mount/remount /x 085040b9....0034....4039 satu pukulan, ganti 0034 dengan 0035 (jika Anda tidak beruntung maka gunakan kombinasi lain untuk menemukan dan menambal vrr_do_mount_check_before_mount (lihat di atas) atau cukup terapkan tambalan hanya untuk / sistem (lihat di posting di atas)
> > > > > > > > > 
> > > > > > > > > 
> > > > > > > > > menghadapi bug setelah Anda menambal
> > > > > > > > 
> > > > > > > > 
> > > > > > > > Oke!
> > > > > > > > gsi atau dsu tidak bisa boot / mendekripsi
> > > > > > 
> > > > > > 
> > > > > > apakah kamu bisa melakukannya sebelumnya?
> > > > > > yess gsi google and stock work fine before now only bootloop
> > > > 
> > > > 
> > > > so my patches worked on stock rom, but bootloops on gsi, right?
> > > > magisk SUU worked on gsi?
> > > 
> > > 
> > > yes stock rom fine but dsu slot bootloop all rom
> > 
> > 
> > just flash gsi into system partition xD
> 
> i don't know why keymaster cannot decrypt userdata

for tests you should 
1. patch boot with stable magisk 24.3+ and try to boot
2. apply my SU patches and try to boot
3. apply "mount /system" patch (https://github.com/topjohnwu/Magisk/issues/5148#issuecomment-1402831067) and try to boot

**Comment by muhammadrafiasyddiq** [comment]
> > > > > > > > > > > Berhasil menambal kernel yang berbeda, gunakan radare2 untuk menemukan alamat SU: /x .fc.0171....0054....4039.fd.0171....0054 harus dua hit, ganti fc pertama dengan fd
> > > > > > > > > > > Mount/remount /x 085040b9....0034....4039 satu pukulan, ganti 0034 dengan 0035 (jika Anda tidak beruntung maka gunakan kombinasi lain untuk menemukan dan menambal vrr_do_mount_check_before_mount (lihat di atas) atau cukup terapkan tambalan hanya untuk / sistem (lihat di posting di atas)
> > > > > > > > > > 
> > > > > > > > > > 
> > > > > > > > > > menghadapi bug setelah Anda menambal
> > > > > > > > > 
> > > > > > > > > 
> > > > > > > > > Oke!
> > > > > > > > > gsi atau dsu tidak bisa boot / mendekripsi
> > > > > > > 
> > > > > > > 
> > > > > > > apakah kamu bisa melakukannya sebelumnya?
> > > > > > > yess gsi google and stock work fine before now only bootloop
> > > > > 
> > > > > 
> > > > > so my patches worked on stock rom, but bootloops on gsi, right?
> > > > > magisk SUU worked on gsi?
> > > > 
> > > > 
> > > > yes stock rom fine but dsu slot bootloop all rom
> > > 
> > > 
> > > just flash gsi into system partition xD
> > 
> > 
> > i don't know why keymaster cannot decrypt userdata
> 
> for tests you should
> 
> 1. patch boot with stable magisk 24.3+ and try to boot
> 2. apply my SU patches and try to boot
> 3. apply "mount /system" patch ([VIVO/iQOO kernel restictions: "Operation not permitted" when executing `su` or mounting to `/system`, or bootloops if rooted #5148 (comment)](https://github.com/topjohnwu/Magisk/issues/5148#issuecomment-1402831067)) and try to boot
> 4. If everything is fine then bug is caused by my ugly patching of vrr_do_mount_check_before_mount. You can patch it differently.
>    Find 3 instructions
>    /x 085040b9....0034....4039
> 
> In my case it's 085040b9680e003408a04039
> 
> ldr w8, [x0, #0x50] cbz w8, #0x1d0 ldrb w8, [x0, #0x28]
> 
> replace 'cbz w8' with 'b' cbz w8, #0x1d0 -> b #0x1d0
> 
> convert it back
> 
> 085040B97300001408A04039
> 
> Final patch for me 085040b9680e003408a04039 -> 085040B97300001408A04039

share you telegram

**Comment by muhammadrafiasyddiq** [comment]
> > > > > > > > > > > Berhasil menambal kernel yang berbeda, gunakan radare2 untuk menemukan alamat SU: /x .fc.0171....0054....4039.fd.0171....0054 harus dua hit, ganti fc pertama dengan fd
> > > > > > > > > > > Mount/remount /x 085040b9....0034....4039 satu pukulan, ganti 0034 dengan 0035 (jika Anda tidak beruntung maka gunakan kombinasi lain untuk menemukan dan menambal vrr_do_mount_check_before_mount (lihat di atas) atau cukup terapkan tambalan hanya untuk / sistem (lihat di posting di atas)
> > > > > > > > > > 
> > > > > > > > > > 
> > > > > > > > > > menghadapi bug setelah Anda menambal
> > > > > > > > > 
> > > > > > > > > 
> > > > > > > > > Oke!
> > > > > > > > > gsi atau dsu tidak bisa boot / mendekripsi
> > > > > > > 
> > > > > > > 
> > > > > > > apakah kamu bisa melakukannya sebelumnya?
> > > > > > > yess gsi google and stock work fine before now only bootloop
> > > > > 
> > > > > 
> > > > > so my patches worked on stock rom, but bootloops on gsi, right?
> > > > > magisk SUU worked on gsi?
> > > > 
> > > > 
> > > > yes stock rom fine but dsu slot bootloop all rom
> > > 
> > > 
> > > just flash gsi into system partition xD
> > 
> > 
> > i don't know why keymaster cannot decrypt userdata
> 
> for tests you should
> 
> 1. patch boot with stable magisk 24.3+ and try to boot
> 2. apply my SU patches and try to boot
> 3. apply "mount /system" patch ([VIVO/iQOO kernel restictions: "Operation not permitted" when executing `su` or mounting to `/system`, or bootloops if rooted #5148 (comment)](https://github.com/topjohnwu/Magisk/issues/5148#issuecomment-1402831067)) and try to boot
> 4. If everything is fine then bug is caused by my ugly patching of vrr_do_mount_check_before_mount. You can patch it differently.
>    Find 3 instructions
>    /x 085040b9....0034....4039
> 
> In my case it's 085040b9680e003408a04039
> 
> ldr w8, [x0, #0x50] cbz w8, #0x1d0 ldrb w8, [x0, #0x28]
> 
> replace 'cbz w8' with 'b' cbz w8, #0x1d0 -> b #0x1d0
> 
> convert it back
> 
> 085040B97300001408A04039
> 
> Final patch for me 085040b9680e003408a04039 -> 085040B97300001408A04039

![Screenshot_20230718_205608](https://github.com/topjohnwu/Magisk/assets/66067362/598b90c4-9d7b-42b1-a115-35ca237c2743)
i can't found su address

**Comment by AarifZ** [comment]
> Successfully patched different kernel, use radare2 to find address SU: /x .fc.0171....0054....4039.fd.0171....0054 should be two hits, replace first fc with fd
> 
> Mount/remount /x 085040b9....0034....4039 one hit, replace 0034 with 0035 (if you're unlucky then use another combinations to find and patch vrr_do_mount_check_before_mount(see above) or just apply patch only for /system (see in posts above) ![Screenshot_20230527-090244_Termux](https://private-user-images.githubusercontent.com/105647092/241361426-fdeec048-0c0f-4ca5-9624-40704e6077c3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI2MTQwOTEsIm5iZiI6MTcwMjYxMzc5MSwicGF0aCI6Ii8xMDU2NDcwOTIvMjQxMzYxNDI2LWZkZWVjMDQ4LTBjMGYtNGNhNS05NjI0LTQwNzA0ZTYwNzdjMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIxNVQwNDE2MzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ZDBkMGFjN2Q0OTUzZGFlNGMzMjJiYjQ0MDZkNWVhNjVjNzAyOWVkYTA1N2JkNWM2ZjNmZmU3Njg5NDY3ZmI3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.FfKN6XgRia1pO5sw5o6bZoYAxwYDkF-hA35loiVSSIw) ![Screenshot_20230527-090109_Termux](https://private-user-images.githubusercontent.com/105647092/241361432-57c5e8a0-2541-43b7-be3a-d9446ec719f8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDI2MTQwOTEsIm5iZiI6MTcwMjYxMzc5MSwicGF0aCI6Ii8xMDU2NDcwOTIvMjQxMzYxNDMyLTU3YzVlOGEwLTI1NDEtNDNiNy1iZTNhLWQ5NDQ2ZWM3MTlmOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBSVdOSllBWDRDU1ZFSDUzQSUyRjIwMjMxMjE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDIzMTIxNVQwNDE2MzFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mNjRlNjU5MmU2OTc1MTcxOGE3NzkyYzk3Nzg0MzVkODg5MzA3NWEyMzE2ZTcwNDU0NWI4NTBiNTY1YmJjOTJlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.rywwiKw2tGwWWGM5rTmRAI-MTyIKlhmcVT109DtGrKY)

## Thank you very much this helped a noob like me to get magisk-26.4 with su working on my Vivo v15
- I've patched /x .fc.0171....0054....4039.fd.0171....0054 (2 hits) 
- /x 085040b9....0034....4039 (1 hit)
- and mount /system by [/system -> /syswxl] 
using stock boot image and then magisk patched it.

My Notes: 
- `Failed to boot when I used Hex edtior pro app for kernel its better to use magisk's hexpatch kernel also stock recommend.`
- `Magisk_patched didn't boot maybe I left vbmeta checked or something on it.`
- `Limitation is that even though its rooted some binaries cant be executed unless selinux is set permissive or moved to system and perm_set 0 2000 0755 u:object_r:system_file:s0 example for frida or similar tools`

[/system -> /syswxl]: https://github.com/topjohnwu/Magisk/issues/5148#issuecomment-1402831067

**Comment by romanovj** [comment]
@AarifZ actually it's possible to patch kernel with radare2, I'm just to lazzy to write how to and to lazzy to write script for automation

`oo+`

then it's become possible to write
https://book.rada.re/basic_commands/write.html

**Comment by AarifZ** [comment]
@romanovj ah seems good maybe someone can fuse the magisk tools and radare2 and some script to make a Boot.img patcher for these devices and throw exception for unsupported kernels. only automation I know is tasker unfortunately possible but not feasible here.
I thought vivo will be as easy as any oppo but after Preloader auth these kernel restrictions lol spent 2 days trying everything ported Twrp tried GSI etc.,, nothing helps everything Bootloops or Twrp fails mounting many partitions , after browsing across xda found magisk suu it worked so after reading the thread few times and following your guide was able to root it definitively. cant thank you enough and everyone's contributions to this thread.

**Comment by Adeyoelabbb** [comment]
Thank you everyone, can you please easy way to access su after rooting samsung galaxy via magisk?
Thanks

**Comment by K11MCH1** [comment]
> @AarifZ actually it's possible to patch kernel with radare2, I'm just to lazzy to write how to and to lazzy to write script for automation
> 
> `oo+`
> 
> then it's become possible to write https://book.rada.re/basic_commands/seeking.html https://book.rada.re/basic_commands/write.html

Is this still for devices with released kernel source, or can be used with closed source?
Asking cause I'm iQOO 11 owner.

**Comment by romanovj** [comment]
> > @AarifZ actually it's possible to patch kernel with radare2, I'm just to lazzy to write how to and to lazzy to write script for automation
> > `oo+`
> > then it's become possible to write https://book.rada.re/basic_commands/seeking.html https://book.rada.re/basic_commands/write.html
> 
> Is this still for devices with released kernel source, or can be used with closed source? Asking cause I'm iQOO 11 owner.

you don't need sources for hexpatching

**Comment by dxkf177** [comment]
> Device: iQOO Z1/VIVO X70 PRO/iQOO NEO3, and more
> Android version: N/A
> Magisk version name: N/A
> Magisk version code: N/A, but more serious since 24302
> 
> Magisk fails to mount /system on some Vivo/iQOO devices.
> ```
> 01-01 08:17:34.721  root   650   652 E Magisk  : mount tmpfs->/system/bin failed with 1: Operation not permitted
> 01-01 08:17:34.721  root   650   652 E Magisk  : mount /dev/Qnh/.magisk/mirror/system/bin/AudioSetParam->/system/bin/AudioSetParam failed with 1: Operation not permitted
> 01-01 08:17:34.721  root   650   652 E Magisk  : mount /dev/Qnh/.magisk/mirror/system/bin/TEST_IServiceManager->/system/bin/TEST_IServiceManager failed with 1: Operation not permitted
> 01-01 08:17:34.721  root   650   652 E Magisk  : mount /dev/Qnh/.magisk/mirror/system/bin/abb->/system/bin/abb failed with 1: Operation not permitted
> ```
> According to a bug report from a "Vivo internal staff" who has a device running userdebug build, `adb root` and manually mount /system works but fails when mounting in Magisk superuser environment. After debugging with him/her/them, we found that the device only allows programs in /system to mount the system partition (move binary that will mount /system to other partitions, and it fails even in adb root environment). I believe that the device has an allowlist and refuses to mount key partitions for programs if their partition is not in the allowlist.
> 
> I think the following two solutions may fix it:
> 1. Reimplement magic mounting logic in rc script language and add it to init.rc.
> 2. Inject Magisk binaries to /system in magiskinit.
> 
> I'm unsure when this problem will be solved, so I opened an issue to track the relevant information.
> 
> 
> ---
> 
> Edit: even successfully mount `su` to `/system/bin/su`, executing `su` will throw an operation not permitted error.

**Comment by AarifZ** [comment]
> > Device: iQOO Z1/VIVO X70 PRO/iQOO NEO3, and more
> > Android version: N/A
> > Magisk version name: N/A
> > Magisk version code: N/A, but more serious since 24302
> > Magisk fails to mount /system on some Vivo/iQOO devices.
> > ```
> > 01-01 08:17:34.721  root   650   652 E Magisk  : mount tmpfs->/system/bin failed with 1: Operation not permitted
> > 01-01 08:17:34.721  root   650   652 E Magisk  : mount /dev/Qnh/.magisk/mirror/system/bin/AudioSetParam->/system/bin/AudioSetParam failed with 1: Operation not permitted
> > 01-01 08:17:34.721  root   650   652 E Magisk  : mount /dev/Qnh/.magisk/mirror/system/bin/TEST_IServiceManager->/system/bin/TEST_IServiceManager failed with 1: Operation not permitted
> > 01-01 08:17:34.721  root   650   652 E Magisk  : mount /dev/Qnh/.magisk/mirror/system/bin/abb->/system/bin/abb failed with 1: Operation not permitted
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > According to a bug report from a "Vivo internal staff" who has a device running userdebug build, `adb root` and manually mount /system works but fails when mounting in Magisk superuser environment. After debugging with him/her/them, we found that the device only allows programs in /system to mount the system partition (move binary that will mount /system to other partitions, and it fails even in adb root environment). I believe that the device has an allowlist and refuses to mount key partitions for programs if their partition is not in the allowlist.
> > I think the following two solutions may fix it:
> > 
> > 1. Reimplement magic mounting logic in rc script language and add it to init.rc.
> > 2. Inject Magisk binaries to /system in magiskinit.
> > 
> > I'm unsure when this problem will be solved, so I opened an issue to track the relevant information.
> > 
> > Edit: even successfully mount `su` to `/system/bin/su`, executing `su` will throw an operation not permitted error.

If done hexpatch installing selinux permissiver in magisk help fix many module issues which aren't executing or having issues.

**Comment by depesh1977** [comment]
I have Y81,android 8.1 on hand. I faced the same situation.
`/x .fc.0171....0054....4039.fd.0171....0054`
It doesn't find anything.Maybe there should be another mask there.
[kernel](https://disk.yandex.ru/d/N-JcZjFodqpbgg) posted to the cloud there

**Comment by AarifZ** [comment]
Have you tried mtksu? @depesh1977

**Comment by depesh1977** [comment]
> Have you tried mtksu? @depesh1977 

No

**Comment by 1q23lyc45** [comment]
Vivo Y52s 5G (PD2057C) FuntouchOS 10 MTK6853
Using MagickBoot to patch the kernel resulted in a bootloop. So I used Radare2 to directly modify it, and the modification was successful. Both Stock ROM and GSI can boot successfully.

0x73 > 0x77 (su > wu)
First, Enter this command to open the kernel using radare2 : `r2 kernel`
Second,
```
[0x00000000]> /x .fc.0171....0054....4039.fd.0171....0054
```
There should be several hits. For example:
```
[0x00000000]> /x .fc.0171....0054....4039.fd.0171....0054
0x003e2e7c hit0_0 3fcd017141010054080540391fd50171e1000054
0x00674bb0 hit0_1 3fcd0171c1000054090540393fd5017161000054
0x00674c3c hit0_2 3fcd017161010054080540391fd5017101010054
[0x00000000]>
```
Write down the hexadecimal digits in the first column. Third, run `oo+`.

Then, run `s <The first hexadecimal digit> ; pd 2`, for example, `s 0x003e2e7c;pd 2`, At this point, the first line contains assembly code starting with cmp, such as `cmp w9, 0x73`. Copy it to clipboard. Enter "wa" then paste, and edit 0x73 to 0x77, and press enter.

Finally, Perform the same operation on the other hexadecimal addresses noted above, then enter q to exit.

Bug: Magisk Daemon still cannot boot up and can only be manually started using the `adb root;adb shell /sbin/magisk --daemon`command.

**Comment by allisonnew359-sys** [comment]
> Device: iQOO Z1/VIVO X70 PRO/iQOO NEO3, and more Android version: N/A Magisk version name: N/A Magisk version code: N/A, but more serious since 24302
> 
> Magisk fails to mount /system on some Vivo/iQOO devices.
> 
> ```
> 01-01 08:17:34.721  root   650   652 E Magisk  : mount tmpfs->/system/bin failed with 1: Operation not permitted
> 01-01 08:17:34.721  root   650   652 E Magisk  : mount /dev/Qnh/.magisk/mirror/system/bin/AudioSetParam->/system/bin/AudioSetParam failed with 1: Operation not permitted
> 01-01 08:17:34.721  root   650   652 E Magisk  : mount /dev/Qnh/.magisk/mirror/system/bin/TEST_IServiceManager->/system/bin/TEST_IServiceManager failed with 1: Operation not permitted
> 01-01 08:17:34.721  root   650   652 E Magisk  : mount /dev/Qnh/.magisk/mirror/system/bin/abb->/system/bin/abb failed with 1: Operation not permitted
> ```
> 
> According to a bug report from a "Vivo internal staff" who has a device running userdebug build, `adb root` and manually mount /system works but fails when mounting in Magisk superuser environment. After debugging with him/her/them, we found that the device only allows programs in /system to mount the system partition (move binary that will mount /system to other partitions, and it fails even in adb root environment). I believe that the device has an allowlist and refuses to mount key partitions for programs if their partition is not in the allowlist.
> 
> I think the following two solutions may fix it:
> 
> 1. Reimplement magic mounting logic in rc script language and add it to init.rc.
> 2. Inject Magisk binaries to /system in magiskinit.
> 
> I'm unsure when this problem will be solved, so I opened an issue to track the relevant information.
> 
> Edit: even successfully mount `su` to `/system/bin/su`, executing `su` will throw an operation not permitted error.

**Comment by canyie** [comment]
More context: https://leadroyal.cn/p/2613/

**Comment by romanovj** [comment]
@canyie Writing to boot partitions is also restricted by the kernel.
